### ISSUE TITLE
Potential error(e.g., deadlock) due to the unreleased lock &lock

### ISSUE THREAD

**Issue by anson-lo** [axboe/liburing#378]
Hi, developers:
       It seems the lock `$lock `is not released if the control flow goes to the` break` statement. The fix is to add the unlock statement before the `break`. Thank you for your checking! It is in the test folder but still worth fixing for higher quality.

The suggested fix is to add `pthread_mutex_unlock(&lock);` before breaking and returning.

https://github.com/axboe/liburing/blob/a209abe65befe603311b6fcd861e14d28e3e2f2d/test/ring-leak2.c#L194
https://github.com/axboe/liburing/blob/a209abe65befe603311b6fcd861e14d28e3e2f2d/test/ring-leak2.c#L198

The relevant code is shown here.

```
static void *client_thread(void *arg)
{
	...; // while loop
			if (type == READ) {
				pthread_mutex_lock(&lock); //the lock is here

				if (cqe->res <= 0) {
					// connection closed or error
					shutdown(conn_i.fd, SHUT_RDWR);
				} else {
					break;   //if it is breaked, the lock is not released
				}
				add_socket_pollin(&ring, conn_i.fd);
				pthread_mutex_unlock(&lock);
			} ...;
		}
              ...;
	}
}

```
Best,

**Comment by isilence** [comment]
Hey @anson-lo, looks you did a good job looking for problems in tests. Do you want to send patches or a PR? Would be appreciated.

**Comment by anson-lo** [comment]
@isilence Thank you for your reply. Sure! I would send the patches soon.

**Comment by anson-lo** [comment]
The patch for this issue. [myfixes.patch.zip](https://github.com/axboe/liburing/files/6827452/myfixes.patch.zip)

```
diff -Naur liburing/test/ring-leak2.c liburing-fix/test/ring-leak2.c
--- liburing/test/ring-leak2.c	2021-07-16 10:35:19.000000000 +0800
+++ liburing-fix/test/ring-leak2.c	2021-07-16 10:37:56.000000000 +0800
@@ -197,6 +197,7 @@
 					// connection closed or error
 					shutdown(conn_i.fd, SHUT_RDWR);
 				} else {
+					pthread_mutex_unlock(&lock);
 					break;
 				}
 				add_socket_pollin(&ring, conn_i.fd);
```

**Comment by axboe** [comment]
Can you send that as a real patch or a pull request? Note the usual format of commits in the tree.

**Comment by anson-lo** [comment]
@axboe , OK, I will send a pull request soon. In addition, could you reopen the #380 and #381?  The problem remains unresolved. Thank you indeed.

**Comment by axboe** [comment]
Thanks! You are mistaken on 380 and 381, I just commented in there.

**Comment by anson-lo** [comment]
Fix in #389.
